copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaProjectorTest.java

```diff
@@ -0,0 +1,464 @@
+        try {
+            SchemaProjector.project(optionalWriter, noDefaultValueReader, array);
+            fail("Reader does not provide a default value.");
+        } catch (SchemaProjectorException e) {
+            // expected
+        }
+
+        Schema nonPromotableReader = SchemaBuilder.array(Schema.BOOLEAN_SCHEMA).build();
+        try {
+            SchemaProjector.project(optionalWriter, nonPromotableReader, array);
+            fail("Neither writer type matches reader type nor writer type can be promoted to reader type");
+        } catch (SchemaProjectorException e) {
+            // expected
+        }
+    }
+
+    @Test
+    public void testMapProjection() throws Exception {
+        Schema writer = SchemaBuilder.map(Schema.INT32_SCHEMA, Schema.INT32_SCHEMA).optional().build();
+
+        Schema reader = SchemaBuilder.map(Schema.INT32_SCHEMA, Schema.INT32_SCHEMA).defaultValue(Collections.singletonMap(1, 2)).build();
+        Object projected = SchemaProjector.project(writer, reader, Collections.singletonMap(3, 4));
+        assertEquals(Collections.singletonMap(3, 4), (Map<?, ?>) projected);
+        projected = SchemaProjector.project(writer, reader, null);
+        assertEquals(Collections.singletonMap(1, 2), (Map<?, ?>) projected);
+
+        Schema promotedReader = SchemaBuilder.map(Schema.INT64_SCHEMA, Schema.FLOAT32_SCHEMA).defaultValue(
+                Collections.singletonMap(3L, 4.5F)).build();
+        projected = SchemaProjector.project(writer, promotedReader, Collections.singletonMap(3, 4));
+        assertEquals(Collections.singletonMap(3, 4), (Map<?, ?>) projected);
+        projected = SchemaProjector.project(writer, promotedReader, null);
+        assertEquals(Collections.singletonMap(3L, 4.5F), (Map<?, ?>) projected);
+
+        Schema noDefaultValueReader = SchemaBuilder.map(Schema.INT32_SCHEMA, Schema.INT32_SCHEMA).build();
+        try {
+            SchemaProjector.project(writer, noDefaultValueReader, null);
+            fail("Reader does not provide a default value.");
+        } catch (SchemaProjectorException e) {
+            // expected
+        }
+
+        Schema nonPromotableReader = SchemaBuilder.map(Schema.BOOLEAN_SCHEMA, Schema.STRING_SCHEMA).build();
+        try {
+            SchemaProjector.project(writer, nonPromotableReader, null);
+            fail("Neither writer type matches reader type nor writer type can be promoted to reader type");
+        } catch (SchemaProjectorException e) {
+            // expected
+        }
+    }
+
+    private void verifyOptionalProjection(Schema writer, Type readerType, Object value, Object defaultValue, boolean optional) {
+        Schema reader = null;
+        assert writer.isOptional();
+        switch (readerType) {
+            case INT8:
+                if (optional) {
+                    reader = SchemaBuilder.int8().optional().defaultValue(defaultValue).build();
```

Comment:
You can reduce this code a lot by using `SchemaBuilder.type(readerType)` to handle all these cases instead of using `SchemaBuilder.int8()`, `SchemaBuilder.int16()`, etc.